IRGen: address a TODO in the autolinking

compnerd opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->

#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

Use the target specific directive generation.  This addresses a TODO to use the
same logic as clang.  Unfortunately, the logic in clang is not accessible
outside of it, so replicate the behaviour.

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -781,18 +781,37 @@ static StringRef encodeForceLoadSymbolName(llvm::SmallVectorImpl<char> &buf,
+    TargetLinkerOptionGenerator LOG(Module.getTargetTriple());
+    auto Opt = LOG.encode(linkLib.getName());
+    AutolinkEntries.push_back(llvm::MDNode::get(ctx,
+                                                llvm::MDString::get(ctx, Opt)));
```
80-column limit.

compnerd replied:
Its at 79-columns.  I find this split more readable, but Ill clang-format it none-the-less.

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -781,18 +781,37 @@ static StringRef encodeForceLoadSymbolName(llvm::SmallVectorImpl<char> &buf,
+  TargetLinkerOptionGenerator(const std::string &Triple) : T(Triple) {}
+  std::string encode(StringRef Library) const {
+    if (T.isWindowsMSVCEnvironment() || T.isWindowsItaniumEnvironment())
+      return "/DEFAULTLIB:" +
```
Nitpick: can you use a SmallString here? I know there won't be too many of these options but still.

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -781,18 +781,37 @@ static StringRef encodeForceLoadSymbolName(llvm::SmallVectorImpl<char> &buf,
-    buf += linkLib.getName();
-    auto flag = llvm::MDString::get(ctx, buf);
-    AutolinkEntries.push_back(llvm::MDNode::get(ctx, flag));
+    TargetLinkerOptionGenerator LOG(Module.getTargetTriple());
```
Might as well use the already-parsed triple from the LLVMContext.

compnerd replied:
I don't see how to get that?  I don't see a `llvm::Triple` being stored anywhere in `LLVMContext`.

jrose-apple replied:
Oops, right, it's in the Swift IRGenModule (`this`) rather than the context. It's just `Triple`.

compnerd replied:
Nice, I didnt even think of looking in the swift IGM.  Good catch!

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -781,18 +781,37 @@ static StringRef encodeForceLoadSymbolName(llvm::SmallVectorImpl<char> &buf,
 }
 
+class TargetLinkerOptionGenerator {
+  static std::string Quoted(StringRef Name) {
```
Lowercase for function names, please (if you keep this).

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -781,18 +781,37 @@ static StringRef encodeForceLoadSymbolName(llvm::SmallVectorImpl<char> &buf,
+      return "/DEFAULTLIB:" +
+             Quoted(Library.str() +
+                    (Library.endswith_lower(".lib") ? "" : ".lib"));
+    else if (T.isPS4())
```
No "else" needed, since the previous "if" ends in a return.

jrose-apple commented on test/IRGen/autolink-coff.swift:
```diff
@@ -15,3 +24,10 @@ import module
+
+// CHECK-MSVC-ASM: .section .drectve
+// CHECK-MSVC-ASM: .ascii " /DEFAULTLIB:module.lib"
+
```
Tests for PS4? (If it's not possible yet, please save for a follow-up commit.)

compnerd replied:
I can even upload them!  I already have them ready to go, just waiting on the PS4 support to be merged :-).  I can remove this handling and make it part of the PS4 target support if you like.

jrose-apple replied:
Up to you: if you think the PS4 patch is going in soon, you can fold it in here; otherwise I'd prefer to make it separate.

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -781,18 +781,38 @@ static StringRef encodeForceLoadSymbolName(llvm::SmallVectorImpl<char> &buf,
 }
 
+class TargetLinkerOptionGenerator {
+  static Twine quoted_import_library(StringRef library) {
```
This is a really unsafe use of Twine. I see that it might be correct in this case, but I don't want to introduce such a trap into the code. Please just pass a string buffer and write to it, like elsewhere in LLVM.

Also, lowerCamelCase please. (We follow the LLVM conventions for everything except variables, where we have both lowerCamelCase and upperCamelCase.)

compnerd commented:
@jrose-apple looking back at the implementation, it seemed overkill.  I simplified it and adjusted the twine usage.  Mind having a look?  Im not at my normal machine right now, so I will test it as soon as I can.  I just want to make sure the overall flow looks reasonable enough to you.

jrose-apple commented:
I'm still not comfortable with this use of Twine. From the [doc comment](http://llvm.org/docs/doxygen/html/classllvm_1_1Twine.html#details):

> A Twine is not intended for use directly and should not be stored, its implementation relies on the ability to store pointers to temporary stack objects which may be deallocated at the end of a statement. Twines should only be used accepted as const references in arguments, when an API wishes to accept possibly-concatenated strings.

I'd like to keep the rules simple so that we don't have people getting the wrong impression about what Twine can and can't do.

compnerd commented:
@jrose-apple as you wish.  Updated to not use twines with references to constant data.

jrose-apple commented:
Thanks, Saleem!

@swift-ci Please test

compnerd commented:
@jrose-apple okay to merge?

compnerd commented:
Merging this since Id like to get the PS4 changes in as well.  I believe all of @jrose-apple's comments have been addressed.

jrose-apple commented:
Yep, sorry to leave you hanging. (And welcome to being a commiter. :-) )
